src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -312,125 +423,55 @@ impl Net {
     // rate limiting budget.
     // Returns true on successful frame delivery.
     fn rate_limited_rx_single_frame(&mut self) -> bool {
```

Comment:
This will not work well. If we limit amount of buffer space we give to `readv` with what we have rate limit for, it will make a lot of packets to be split into multiple `readv` calls. Example: we have 32K allowance from rate limiter. We give it to the `readv`. TAP has 64K packet ready to be read. `readv` happily writes first 32K into provided buffer. We have not idea that this is not the whole packet, so we process it as a whole packet. Next time we call `readv` it will write remaining part of the packet, but again, we have no idea that it is not a new packet.
Basically because we have not way of asking TAP about how big is the pending packet, we cannot rate limit before `readv`.

Reply:
The problem with rate limiting before is that we don't know how much we should rate limit. We only know that after we read from the TAP. The next frame in the TAP might be 52 bytes, so if we were to rate limit by the size of the available buffer we would penalize the guest.

Unfortunately, we can't know how many bytes we can read out of the TAP device (If we did, this whole PR would be significantly simpler :))

Now a ping is not the usual case here, but (1) right now the buffers we have are ~72K big and (2) even with mergeable buffers merged, we would be rate-limiting in chunks of an MTU which might penalize traffic that consists of many small frames. I've not idea whether such a scenario exists realistically, but I would be unwilling to change things before understanding whether that's a valid use-case.

Reply:
without mrg_rxbuf we check the capacity of the desc chain to be 64K minimum. So we always pass 64K to readv. With mrg_rxbuf we will do approximately the same thing. So we should never have 1 packet split between 2 `readv` calls

Reply:
Ack, makes sense, thanks for the explanation!